experimental/dds/tree2/src/class-tree/toFlexSchema.ts

```diff
@@ -0,0 +1,246 @@
+): FlexTreeFieldSchema {
+	let kind: FlexFieldKind;
+	let types: ImplicitAllowedTypes;
+	if (schema instanceof FieldSchema) {
+		kind = convertFieldKind.get(schema.kind) ?? fail("Invalid field kind");
+		types = schema.allowedTypes;
+	} else {
+		kind = FieldKinds.required;
+		types = schema;
+	}
+	const allowedTypes = convertAllowedTypes(schemaMap, types);
+	return FlexTreeFieldSchema.create(kind, allowedTypes);
+}
+
+const convertFieldKind = new Map<FieldKind, FlexFieldKind>([
+	[FieldKind.Optional, FieldKinds.optional],
+	[FieldKind.Required, FieldKinds.required],
+]);
+
+/**
+ * Normalizes an {@link ImplicitAllowedTypes} into  {@link AllowedTypes}.
+ */
+export function convertAllowedTypes(
+	schemaMap: Map<TreeNodeSchemaIdentifier, () => FlexTreeNodeSchema>,
+	schema: ImplicitAllowedTypes,
+): FlexAllowedTypes {
+	if (isReadonlyArray(schema)) {
+		return normalizeFlexListEager(schema).map((item) => convertNodeSchema(schemaMap, item));
+	}
+	return [convertNodeSchema(schemaMap, schema)];
+}
+
+const builder = { name: "simple schema" };
+
+/**
+ * Normalizes an {@link ImplicitAllowedTypes} into {@link AllowedTypes}.
+ */
+export function convertNodeSchema(
+	schemaMap: Map<TreeNodeSchemaIdentifier, () => FlexTreeNodeSchema>,
+	schema: TreeNodeSchema,
+): () => FlexTreeNodeSchema {
+	// TODO: nice error if multiple schema have the same identifier
+	const final = getOrCreate(schemaMap, schema.identifier, () => () => {
+		let out: FlexTreeNodeSchema;
+		const kind = schema.kind;
+		switch (kind) {
+			case NodeKind.Leaf: {
+				const cached =
+					cachedFlexSchemaFromClassSchema(schema) ??
+					fail("leaf schema should be pre-cached");
+				assert(schemaIsLeaf(cached), "expected leaf");
+				return cached;
+			}
+			case NodeKind.Map: {
+				const fieldInfo = schema.info as ImplicitAllowedTypes;
+				const field = FlexTreeFieldSchema.create(
+					FieldKinds.optional,
+					convertAllowedTypes(schemaMap, fieldInfo),
+				);
+				// Lookup of cached schema is done here instead of before since walking the schema recursively to populate schemaMap is still required.
+				const cached = cachedFlexSchemaFromClassSchema(schema);
+				out = cached ?? FlexMapNodeSchema.create(builder, brand(schema.identifier), field);
+				break;
+			}
+			case NodeKind.List: {
+				const fieldInfo = schema.info as ImplicitAllowedTypes;
+				const field = FlexTreeFieldSchema.create(
+					FieldKinds.sequence,
+					convertAllowedTypes(schemaMap, fieldInfo),
+				);
+				const cached = cachedFlexSchemaFromClassSchema(schema);
+				out =
+					cached ?? FlexFieldNodeSchema.create(builder, brand(schema.identifier), field);
+				break;
+			}
+			case NodeKind.Object: {
+				const info = schema.info as Record<string, ImplicitFieldSchema>;
+				const fields: Record<string, FlexTreeFieldSchema> = Object.create(null);
+				for (const [key, value] of Object.entries(info)) {
+					// This code has to be careful to avoid assigned to __proto__ or similar built in fields.
```

Comment:
Use of Object.defineProperty avoids assigning to built-in inherited fields. Using = here would be the bad thing.

Reply:
A comment like the following would be clearer:

// Using 'Object.defineProperty(..)' rather than 'fields[key]' avoids inadvertently setting the prototype if the user declares a field named  "__proto__".